vehicles/Hero/main.cc

```diff
@@ -0,0 +1,828 @@
+                                             .cb_size = 0,
+                                             .stack_mem = nullptr,
+                                             .stack_size = 256 * 4,
+                                             .priority = (osPriority_t)osPriorityNormal,
+                                             .tz_module = 0,
+                                             .reserved = 0};
+osThreadId_t chassisTaskHandle;
+// TODO:
+// 1. test keyboard and mouse data
+// 2. test relay
+
+// Params Initialization
+static control::MotorCANBase* fl_motor = nullptr;
+static control::MotorCANBase* fr_motor = nullptr;
+static control::MotorCANBase* bl_motor = nullptr;
+static control::MotorCANBase* br_motor = nullptr;
+static control::Chassis* chassis = nullptr;
+
+void chassisTask(void* arg) {
+  UNUSED(arg);
+  // motors initialization
+  control::MotorCANBase* motors_can1_chassis[] = {fl_motor, fr_motor, bl_motor, br_motor};
+  control::Motor4310* motor_can1_gimbal[] = {yaw_motor};
+  float vx_keyboard = 0, vy_keyboard = 0;
+  float vx_remote = 0, vy_remote = 0;
+  float vx = 0, vy = 0, wz = 0;
+
+  // waiting for the start signal
+  while (true) {
+    if (dbus->keyboard.bit.B || dbus->swr == remote::DOWN) break;
+    osDelay(100);
+  }
+
+  while (true) {
+
+    // Elevation(shift)
+    ElevationDetect.input(dbus->keyboard.bit.SHIFT);
+    if (ElevationDetect.posEdge()) {
+      Elevation = !Elevation;
+      if (Elevation) {
+        relay->On();
+      } else {
+        relay->Off();
+      }
+      osDelay(100);
+    }
+
+    // read the data from keyboard
+    if (dbus->keyboard.bit.A) vx_keyboard -= 61.5;
+    if (dbus->keyboard.bit.D) vx_keyboard += 61.5;
+    if (dbus->keyboard.bit.W) vy_keyboard += 61.5;
+    if (dbus->keyboard.bit.S) vy_keyboard -= 61.5;
+    // process the data from keyboard
+    if (-35 <= vx_keyboard && vx_keyboard <= 35) vx_keyboard = 0;
+    if (-35 <= vy_keyboard && vy_keyboard <= 35) vy_keyboard = 0;
+    // soft start and stop in the keyboard
+    if (vx_keyboard > 0)
+      vx_keyboard -= 60;
+    else if (vx_keyboard < 0)
+      vx_keyboard += 60;
+    if (vy_keyboard > 0)
+      vy_keyboard -= 60;
+    else if (vy_keyboard < 0)
+      vy_keyboard += 60;
+    // clip the keyboard data
+    vx_keyboard = clip<float>(vx_keyboard, -1200, 1200);
+    vy_keyboard = clip<float>(vy_keyboard, -1200, 1200);
+    // read the data from the remote controller
+    vx_remote = dbus->ch0;
+    vy_remote = dbus->ch1;
+    // sum whole x and y direction data
+    vx = vx_keyboard + vx_remote;
+    vy = vy_keyboard + vy_remote;
+    // rotation velocity calculation
+    wz = yaw_sum * 120000; // 120000 is magic number
+
+    // Update the speed by power limit from referee
+    if (!Dead && !Elevation) {
+      chassis->SetSpeed(vx, vy, wz);
+      yaw_motor->SetOutput(0, 0, 8, 1.5, 0);
+      control::Motor4310::TransmitOutput(motor_can1_gimbal, 1);
+    } else {
+      chassis->SetSpeed(0, 0, 0);
+    }
```

Comment:
Not sure what elevation does, better explain more and the numbers in it with comments

Reply:
elevation is for the gimbal. when elevation, the mouse is used to control the shooter rather than chassis.